lib/database/migrations/20240424092020-add-qc-flag-effective-period-columns.js

```diff
@@ -0,0 +1,244 @@
+/**
+ * @license
+ * Copyright CERN and copyright holders of ALICE O2. This software is
+ * distributed under the terms of the GNU General Public License v3 (GPL
+ * Version 3), copied verbatim in the file "COPYING".
+ *
+ * See http://alice-o2.web.cern.ch/license for full licensing information.
+ *
+ * In applying this license CERN does not waive the privileges and immunities
+ * granted to it by virtue of its status as an Intergovernmental Organization
+ * or submit itself to any jurisdiction.
+ */
+
+const { timestampToMysql } = require('../../server/utilities/timestampToMysql');
+
+/**
+ * Discard effective periods of flags belonging to given scope
+ * @param {QueryInterface} queryInterface sequelize interface
+ * @param {Period} period period which discards other effective periods
+ * @param {object} scope scope of QC flags
+ * @param {object} scope.runNumber run number
+ * @param {object} scope.dplDetectorId dpl detector id
+ * @param {object} [scope.dataPassId] data pass id - exclusive with simulationPassId
+ * @param {object} [scope.simulationPassId] simulation pass id - exclusive with dataPassId
+ * @param {object} scope.createdAt creation timestamp of discarding flag,
+ * only effective periods of QC flags created before this timestamp will be updated
+ * @return {Promise<void>} promise
+ */
+const _updateEffectivePeriodForScopeAndPeriod = async (
+    queryInterface,
+    { from: intersectingPeriodFrom, to: intersectingPeriodTo },
+    { runNumber, dplDetectorId, dataPassId, simulationPassId, createdAt },
+    { transaction },
+) => {
+    const joinClause =
+    `
+        INNER JOIN quality_control_flags as qcf
+            ON qcfep.flag_id = qcf.id
+                AND qcf.run_number = ${runNumber} 
+                AND qcf.dpl_detector_id = ${dplDetectorId}
+                AND unix_timestamp(qcf.created_at) * 1000 < ${createdAt}
+        INNER JOIN ${
+    dataPassId !== undefined
+        ? `data_pass_quality_control_flag as dpqcf
+                    ON dpqcf.quality_control_flag_id = qcf.id AND dpqcf.data_pass_id = ${dataPassId}
+                    `
+        : `simulation_pass_quality_control_flag as spqcf
+                    ON spqcf.quality_control_flag_id = qcf.id AND spqcf.simulation_pass_id = ${simulationPassId}
+                    `
+}
+    `;
```

Comment:
You should never have nested template string, it is very hard to read. Basically, what I would do here is:
```js
let joinClause = `INNER JOIN ...` // common part
if (dataPassId !== undefined) {
    joinClause += `\nINNER JOIN data_pass_quality_control_flag as dpqcf
                    ON dpqcf.quality_control_flag_id = qcf.id AND dpqcf.data_pass_id = ${dataPassId}`;
}
if (simulationPassId !== undefined) {
    joinClause += `\nINNER JOINsimulation_pass_quality_control_flag as spqcf
                    ON spqcf.quality_control_flag_id = qcf.id AND spqcf.simulation_pass_id = ${simulationPassId}`
}
```

Reply:
applied